src/System.Net.Http/tests/FunctionalTests/HttpContentTest.cs

```diff
@@ -430,6 +434,55 @@ public async Task ReadAsStringAsync_SetNoCharset_DefaultCharsetUsed()
             Assert.Equal(sourceString, result);
         }
 
+        [Fact]
+        public async Task ReadAsStringAsync_SetQuotedCharset_ParsesContent()
+        {
+            string sourceString = "some string";
+            byte[] contentBytes = Encoding.UTF8.GetBytes(sourceString);
+
+            var content = new MockContent(contentBytes);
+            content.Headers.ContentType = new MediaTypeHeaderValue("text/plain");
+            content.Headers.ContentType.CharSet = "\"utf-8\"";
+
+            string result = await content.ReadAsStringAsync();
+
+            Assert.Equal(sourceString, result);
+        }
+
+        [Theory]
+        [InlineData("\"\"invalid")]
+        [InlineData("invalid\"\"")]
+        [InlineData("\"\"invalid\"\"")]
+        [InlineData("\"invalid")]
+        [InlineData("invalid\"")]
+        public async Task ReadAsStringAsync_SetInvalidContentTypeHeader_DefaultCharsetUsed(string charset)
```

Comment:
This test is set up differently, because we actually can't instantiate `MediaTypeHeaderValue` with any of these strings.

Reply:
>This test is set up differently, because we actually can't instantiate MediaTypeHeaderValue with any of these strings.

Can't you just use TryAddWithoutValidation() method to add the improperly formatted headers?  Then use the .ReadAsStringAsync() to validate what you end up parsing for charset?

Having an end-to-end test with loopback server is ok.  But you could also try this approach as well.

Reply:
This approach is way cleaner. Thanks.